hoodie-hive/src/main/java/com/uber/hoodie/hive/HiveSyncTool.java

```diff
@@ -19,64 +19,161 @@
+import java.util.stream.Collectors;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat;
+import org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import parquet.schema.MessageType;
 
 /**
- * Tool to sync new data from commits, into Hive in terms of
+ * Tool to sync a hoodie HDFS dataset with a hive metastore table.
+ * Either use it as a api HiveSyncTool.syncHoodieTable(HiveSyncConfig)
+ * or as a command line java -cp hoodie-hive.jar HiveSyncTool [args]
  *
- *  - New table/partitions
- *  - Updated schema for table/partitions
+ * This utility will get the schema from the latest commit and will sync hive table schema
+ * Also this will sync the partitions incrementally
+ * (all the partitions modified since the last commit)
  */
+@SuppressWarnings("WeakerAccess")
 public class HiveSyncTool {
 
+  private static Logger LOG = LoggerFactory.getLogger(HiveSyncTool.class);
+  private final HoodieHiveClient hoodieHiveClient;
+  private final HiveSyncConfig cfg;
+
+  public HiveSyncTool(HiveSyncConfig cfg, HiveConf configuration, FileSystem fs) {
+    this.hoodieHiveClient = new HoodieHiveClient(cfg, configuration, fs);
+    this.cfg = cfg;
+  }
 
-    /**
-     * Sync to Hive, based on day based partitioning
-     *
-     * @param cfg
-     */
-    public static void sync(HiveSyncConfig cfg) {
-        // Configure to point to which metastore and database to connect to
-        HoodieHiveConfiguration apiConfig =
-                HoodieHiveConfiguration.newBuilder().hadoopConfiguration(new Configuration())
-                        .hivedb(cfg.databaseName)
-                        .hiveJdbcUrl(cfg.jdbcUrl)
-                        .jdbcUsername(cfg.hiveUser)
-                        .jdbcPassword(cfg.hivePass)
-                        .build();
+  public void syncHoodieTable() {
+    LOG.info("Trying to sync hoodie table" + cfg.tableName + " with base path " + hoodieHiveClient
+        .getBasePath() + " of type " + hoodieHiveClient
+        .getTableType());
+    // Check if the necessary table exists
+    boolean tableExists = hoodieHiveClient.doesTableExist();
+    // Get the parquet schema for this dataset looking at the latest commit
+    MessageType schema = hoodieHiveClient.getDataSchema();
+    // Sync schema if needed
+    syncSchema(tableExists, schema);
 
-        HoodieDatasetReference datasetReference =
-                new HoodieDatasetReference(cfg.tableName, cfg.basePath, cfg.databaseName);
+    LOG.info("Schema sync complete. Syncing partitions for " + cfg.tableName);
+    // Get the last time we successfully synced partitions
+    Optional<String> lastCommitTimeSynced = Optional.empty();
+    if (tableExists) {
+      lastCommitTimeSynced = hoodieHiveClient.getLastCommitTimeSynced();
```

Comment:
👍  x 💯  . (if only we can make the cleaner also do this & every other meta operation, HDFS can scale a ton more)